scheduler/scheduler.go

```diff
@@ -187,9 +198,59 @@ func (driver *MesosSchedulerDriver) init() error {
 	driver.messenger.Install(driver.slaveLost, &mesos.LostSlaveMessage{})
 	driver.messenger.Install(driver.frameworkMessageRcvd, &mesos.ExecutorToFrameworkMessage{})
 	driver.messenger.Install(driver.frameworkErrorRcvd, &mesos.FrameworkErrorMessage{})
+
+	driver.masterDetector.Detect(detector.AsMasterChanged(driver.masterDetected))
 	return nil
 }
 
+// lead master detection callback.
+func (driver *MesosSchedulerDriver) masterDetected(master *mesos.MasterInfo) {
+	if driver.Status() == mesos.Status_DRIVER_ABORTED {
+		log.Info("Ignoring master change because the driver is aborted.")
+		return
+	}
+
+	// Reconnect every time a master is dected.
+	if driver.Connected() {
+		log.V(3).Info("Disconnecting scheduler.")
+		driver.Scheduler.Disconnected(driver)
+	}
+
+	driver.setConnected(false)
+
+	if master != nil {
+		log.Infof("New master %s detected\n", master.String())
+
+		pid, err := upid.Parse(master.GetPid())
+		if err != nil {
+			panic("Unable to parse Master's PID value.") // this should not happen.
+		}
+
+		driver.MasterPid = pid // save for downstream ops.
+
+		if driver.credential != nil {
+			if err := driver.authenticate(); err != nil {
```

Comment:
authentication should probably not be done synchronously in the detector handler since it could potentially block and we really should try to avoid blocking for long in callbacks like this.